### ISSUE TITLE
Are there any examples for constraints that depends on input?

### ISSUE THREAD

**Issue by ArnaudDeza** [andruekonst/ConstraiNet#1]
Hi,

I am interested in your work and I was wondering if you could point me to an example where ConstraiNet is used for a constraint of the neural network that is based on the input to the network.

I want to use your method to ensure a constraint Ax<=b is met, however, the A matrix depends on the input whereas b is fixed.

Thanks,
Arnaud

**Comment by knochen336** [comment]
Hello, have you solved this problem?
I would also like to use a similar approach to satisfy the constraint Ax<=b, although it is different from your goal. My b matrix depends on the input, while A is fixed.

**Comment by andruekonst** [comment]
Hello,

The input-dependent constraints are not currently implemented directly inside the `ConstraiNetLayer` wrapper, however the underlying implementations can be instantiated at a forward pass with dynamic `A` and `b` tensors.
It means that you can manually construct an implementation module, for example `CenterDiameterBasedNN` in case of linear constraints, when the mode is `"ray_shift"`, or `ProjectionCenterDiameterNN` when the mode is `"center_projection"`. These modules take `A, b` as input at initialization, and then generate feasible output at `forward(z)` call. For example:

```python
z = ...  # some input feature vector
A = generate_dynamic_A(z)  # generate constraints left-hand side matrix
b = generate_dynamic_b(z)  # generate constraints right-hand side vector
point = find_static_feasible_point()  # find a point that is feasible for every generated `(A, b)` pair
# point = find_dynamic_feasible_point(A, b)  # if you have some analytical way to infer a feasible point from `z` or `(A, b)`

x = ProjectionCenterDiameterNN(A, b, point).forward(z)  # x satisfy constraints `(A, b)`
```

If you then call `some_loss(x).backward()`, gradients will be available for `A, b` too, and will be propagated through them to `z`.
If this behavior is not desired, `ProjectionCenterDiameterNN(A.detach(), b.detach(), point).forward(z)` can be used.

Please note that the initializations of `CenterDiameterBasedNN` and `ProjectionCenterDiameterNN` are almost zero-cost: the construction at the inference stage (e.g. `x = ProjectionCenterDiameterNN(A, b, point).forward(z)`) will not perform introduce any additional computations, and will not lead to performance degradation, comparing to the case when `ProjectionCenterDiameterNN` is initialized once (when `A, b` are static, independent on the input `z`).



**The most important part is how to generate a feasible point**. It is the main reason why such functionality currently is not included into the `ConstraiNetLayer ` wrapper. Formally, if we call a feasible set, that depends on the input `z`, as $\Omega_z$, we have to construct some map $p: z \rightarrow \Omega_z$, that determines just one arbitrary feasible point for $\Omega_z$. Please note that **this limitation is caused by linear programming algorithmic complexity**: the problem of finding at least one feasible point (or just answering yes/no question about such point existence) can be very hard in general case. Therefore to be able to generate feasible points inside **dynamic** constraints (i.e. for different systems `A, b`), a user has to provide some way to calculate $p(z)$ in a fast way (e.g. in polynomial time). If not, the only remaining way would be solving linear programming problem at each forward pass, that can be unreasonably slow.
 
For some problems a *universal* feasible point can be found, that satisfy constraints for all `A, b` pairs, so the map is constant: $p = const$ and the constant can be determined in advance. But sometimes such points do not exist. For example, there is definitely no *universal* feasible point if intersection of all possible feasible sets is an empty set. In such a case you will have to create a method for feasible point estimation, based on input `z`.

For example, imagine a 2d square with a unit side (the square interior is a feasible set), that slides to the right, depending on the input vector `z`, such that the square origin coordinates are `(z[1], 0)`: if `z[1]` can take values from large interval, e.g. $[0, 10]$, the squares with origins `(0, 0)` and `(10, 0)` will have an empty intersection. However we can define a function $p(z) = (z_1, 0)$, that will deterministically map $z$ to the feasible set $\Omega_z$.

We will try to improve the library interface to be able to handle dynamic `A, b` matrices, given user-specified $p(z)$ mapping.

**Comment by KoperSloper** [comment]
Hey,

Is it sufficient for a universal point to be feasible, or does it strictly have to be an interior point? I.e. could I use a point on the edge of the set that's always feasible?
